src/Act/Decompile.hs

```diff
@@ -0,0 +1,475 @@
+          pure $ Constructor
+            { _cname = T.unpack cs.name
+            , _cinterface = fst cs.creation
+            , _cpreconditions = nub ps
+            , _cpostconditions = mempty
+            , _invariants = mempty
+            , _initialStorage = updates
+            }
+        _ -> error "Internal Error: mkConstructor called on a non Success branch"
+  | otherwise = Left "TODO: decompile constructors with multiple branches"
+
+  where
+    addDefaults :: Map (Integer, Integer) (Text, SlotType) -> DistinctStore -> DistinctStore
+    addDefaults layout (DistinctStore storage) = 
+      DistinctStore $ Map.foldr (\(name, typ) s -> case Map.lookup name storage of
+                                    Just _ -> s
+                                    Nothing -> Map.insert name (LitInt nowhere 0, typ) s) storage layout
+    
+-- | Build behaviour specs from the summarized bytecode
+mkBehvs :: EVMContract -> Either Text [Behaviour]
+mkBehvs c = concatMapM (\(i, bs) -> mapM (mkbehv i) (Set.toList bs)) (Map.toList c.runtime)
+  where
+    mkbehv :: Method -> EVM.Expr EVM.End -> Either Text Behaviour
+    mkbehv method (EVM.Success props _ retBuf store) = do
+      pres <- flattenProps <$> mapM (fromProp (invertLayout c.storageLayout)) props
+      ret <- case method.output of
+        [] -> Right Nothing
+        [(_, typ)] -> case abiKind typ of
+          Static -> case typ of
+            AbiTupleType _ -> Left "cannot decompile methods that return a tuple"
+            AbiFunctionType -> Left "cannot decompile methods that return a function pointer"
+            _ -> do
+              v <- fromWord (invertLayout c.storageLayout) . Expr.readWord (EVM.Lit 0) $ retBuf
+              pure . Just . TExp SInteger $ v
+          Dynamic -> Left "cannot decompile methods that return dynamically sized types"
+        _ -> Left "cannot decompile methods with multiple return types"
+      rewrites <- case Map.toList store of
+        [(EVM.SymAddr "entrypoint", contract)] -> do
+          partitioned <- decomposeStorage (invertLayout c.storageLayout) contract.storage
+          mkRewrites c.name partitioned
+        [(_, _)] -> error $ "Internal Error: state contains a single entry for an unexpected contract:\n" <> show store
+        [] -> error "Internal Error: unexpected empty state"
+        _ -> Left "cannot decompile methods that update storage on other contracts"
+      pure $ Behaviour
+        { _contract = T.unpack c.name
+        , _interface = behvIface method
+        , _name = T.unpack method.name
+        , _preconditions = nub pres
+        , _caseconditions = mempty -- TODO: what to do here?
+        , _postconditions = mempty
+        , _stateUpdates = rewrites
+        , _returns = ret
+        }
+    mkbehv _ _ = error "Internal Error: mkbehv called on a non Success branch"
+
+-- TODO: we probably need to diff against the prestore or smth to be sound here
```

Comment:
Is this true? I don't really understand this comment.

Reply:
reviewed this section pretty carefully together with @zoep and we think this section is OK.
```suggestion
```